### PR TITLE
Nav menu

### PR DESCRIPTION
**Pull request by juzerzarif**
# Description of Changes

- Changed the slide menu to a temporary mini variant drawer
- Added tooling support for mui tree shaking and ts path aliasing

# Motivation & Context

- Design update
- Better production builds and DX

# Type of change

- [ ] 🐛 Bug fix
- [x] ✨ New feature/Enhancement
- [ ] 🛠 Technical currency
- [x] 💎 Tooling change

# Snapshots

# Discussion:

There's a specific case (for larger screen sizes) for the slide menu that we need to make a decision on:
When the drawer is closed and the user clicks on an icon, right now that just opens the drawer. This was done because if the icon is on an expanding row you have to open the drawer to show the user the nested list. However when a user clicks on an icon on a "leaf" row that just leads to a route, the drawer will still open up and they'll just click the same row again. We could just select the "leaf" row instead of opening the drawer but then that creates the problem of users not being able to anticipate the result of their action without knowing the application's functionality. If a user clicks on the mini drawer once and it opens up the drawer and then he clicks on it a second time but this time the drawer doesn't open and he's just routed to a new screen, that's gonna make for confusing UX. The way I see it, we have three options:
a) Keep what we have right now (drawer opens up on every click)
b) Force all top level nav items to have a sub list
c) Scrap the mini variant drawer (which would hurt me a little inside but I'd be fine)

### CODE REVIEW

**Comment by Berkyle** [comment]
That gif is _exactly_ what I was imagining; this looks great. I'd say you can rest easy knowing option (c) is out of the question.

I'm highly partial towards options (a) and (b) - every interaction with the mini variant drawer should pop it out. I see several benefits to proceeding with this design:

- There will be _much_ less pressure to determine the correct material icon should be used for each section of the site, 
   - This is a concern I had while choosing an item for the events page (the calendar icon is rather vague, but there aren't many better options).
- This pattern is both easier to design (for us) and easier to intuit (for the user). We can associate related features much more easily. 
- This option facilitates future optimization if we need it, like lazy loading nav-item-specific networked resources.
- Honestly it just looks good and provides a very seamless, easy-to-navigate experience.

**Comment by Berkyle** [comment]
I'm wrestling a bit with the implications of option (b) - should every nav drawer section have nested nav items? I.e., should j-boards also have a nested dropdown? Benefits might include:

- A more consistent UI.
- It would look silly to for a user to expand a list with only one item, and that forces us to try to fill each nested list and therefore forces us to anticipate what a user might expect from each feature.
   - This could also be a drawback, of course... what if a feature truly has no need of additional options?

This is a consideration that would be difficult to anticipate our need for at this phase, and I don't think it should hold up this PR. **My gut tells me go with (b)**.

File: src/App.tsx [hunk 1]
```diff
@@ -23,13 +21,9 @@ const App: React.FC = () => {
...
     <React.Fragment>
       <CssBaseline />
-      <Router>
-        <Resident51Contexts>
-          <Navigation>
-            <Routes />
-          </Navigation>
-        </Resident51Contexts>
-      </Router>
+      <Resident51Contexts>
```
Comment (Berkyle) [inline src/App.tsx:25]: just to minimize unnecessary changes can we add the Router back outside of `Resident51Contexts`

File: src/components/navigation/NavigationItems.tsx [hunk 2]
```diff
@@ -25,6 +28,12 @@ const navigationItems: NavigationItem[] = [
       },
     ],
   },
+  {
+    id: 'jboards',
+    text: 'Jboards',
+    icon: <GavelIcon />,
+    path: '/path',
+  },
```
Comment (Berkyle) [inline src/components/navigation/NavigationItems.tsx]: to implement option (b) as listed in the PR discussion, this will need to have a nested navigation item. I know it's not created yet but we should probably be proactive about this pattern.

File: src/components/navigation/SlideMenu.tsx [hunk 3]
```diff
@@ -1,156 +1,69 @@
...
   onRequestClose: () => void;
 }
 
 const SlideMenu: React.FC<SlideMenuProps> = props => {
-  const { open, onRequestClose } = props;
+  const { open, onRequestOpen, onRequestClose } = props;
+  const theme = useTheme();
+  const isMobile = useMediaQuery(theme.breakpoints.down('xs'));
   const classes = useStyles();
+  const drawerClass: string = clsx(classes.drawer, {
```
Comment (Berkyle) [inline src/components/navigation/SlideMenu.tsx]: the compiler config is smart enough to infer this type, there is never really a need to specify primitive types
```suggestion
  const drawerClass = clsx(classes.drawer, {
```

File: src/components/navigation/SlideMenu.tsx [hunk 4]
```diff
@@ -1,156 +1,69 @@
...
+    [classes.drawerOpen]: !isMobile && open,
+    [classes.drawerClose]: !isMobile && !open,
+  });
+  const navListClass: string = clsx(classes.navListContainer, { [classes.navListClose]: !open });
```
Comment (Berkyle) [inline src/components/navigation/SlideMenu.tsx]: same comment as above
```suggestion
  const navListClass = clsx(classes.navListContainer, { [classes.navListClose]: !open });
```

File: src/components/navigation/SlideMenuContext.tsx [hunk 5]
```diff
@@ -0,0 +1,75 @@
+import React, { createContext, useCallback, useContext, useEffect, useMemo, useState } from 'react';
```
Comment (Berkyle) [inline src/components/navigation/SlideMenuContext.tsx:1]: for consistency can we move this file to `src/contexts/ui/` and add the provider to the `Resident51Contexts` component

File: src/components/navigation/SlideMenuContext.tsx [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+  const contextValue = useMemo(
+    () => ({
+      expandedNavItems,
+      activeNavItem,
+      expandNavItem: addExpandedNavItem,
+      collapseNavItem: removeExpandedNavItem,
+    }),
+    [activeNavItem, addExpandedNavItem, expandedNavItems, removeExpandedNavItem],
+  );
```
Comment (Berkyle) [inline src/components/navigation/SlideMenuContext.tsx:67]: memoizing a context value is unnecessary if every property is in the dependency array and we re-create a new array on each setState anyway. we would never re-use an old value

File: src/components/navigation/SlideMenu.tsx [hunk 7]
```diff
@@ -1,156 +1,69 @@
-import React, { useMemo, useState } from 'react';
+import React, { ReactNode, useCallback, useMemo } from 'react';
```
Comment (Berkyle) [inline src/components/navigation/SlideMenu.tsx]: looks like we use `React.ReactNode` below, can we remove the import of ReactNode? I also prefer `React.ReactNode` anyway.

Comment (Berkyle) [reply src/components/navigation/SlideMenu.tsx]: why wasn't this caught by our linter? the import plugin usually spots this.

File: src/components/navigation/SlideMenu.tsx [hunk 8]
```diff
@@ -1,156 +1,69 @@
...
+  const theme = useTheme();
+  const isMobile = useMediaQuery(theme.breakpoints.down('xs'));
   const classes = useStyles();
+  const drawerClass: string = clsx(classes.drawer, {
+    [classes.drawerOpen]: !isMobile && open,
+    [classes.drawerClose]: !isMobile && !open,
+  });
+  const navListClass: string = clsx(classes.navListContainer, { [classes.navListClose]: !open });
+
+  const updateRootStyles = useCallback(
```
Comment (Berkyle) [inline src/components/navigation/SlideMenu.tsx]: nit: update name to `updateDrawerZIndex` (if i'm interpreting this right)

File: src/components/navigation/SlideMenuItems.tsx [hunk 9]
```diff
@@ -1,86 +1,123 @@
...
-    });
-  }, [history]);
+  const hasSubList = !!subItemList?.length;
+  const sublistOpen = !!menuState?.expandedNavItems.includes(id);
+  const isSelected: boolean = ((): boolean => {
+    if (menuState?.activeNavItem && menuState?.activeNavItem?.path === path) return true;
+    if (isDrawerOpen) return false;
+    if (menuState?.activeNavItem?.parents.includes(id)) return true;
+    return false;
+  })();
```
Comment (Berkyle) [inline src/components/navigation/SlideMenuItems.tsx]: I appreciate the tendency towards const but this is a bitch and a half to read.  what are your thoughts on the following
```suggestion
  let isSelected = false;
  const activeNav = menuState?.activeNavItem;
  if (activeNav?.path === path) {
    isSelected = true;
  } else if (!isDrawerOpen && activeNav?.parents.includes(id)) {
    isSelected = true;
  }
```

File: src/components/navigation/SlideMenuItems.tsx [hunk 10]
```diff
@@ -1,86 +1,123 @@
...
+    hasSubList,
+    path,
+    isDrawerOpen,
+    sublistOpen,
+    onRequestOpen,
+    menuState,
+    id,
+    history,
+    onRequestClose,
+  ]);
```
Comment (Berkyle) [inline src/components/navigation/SlideMenuItems.tsx]: too many dependency array items to justify a useCallback - the only props this doesn't use are ones I would expect to never change, this isn't worth sacrificing legibility

File: src/components/navigation/SlideMenuItems.tsx [hunk 11]
```diff
@@ -1,86 +1,123 @@
...
 
-  if (props.isVisible && !props.isVisible()) {
+  useEffect(() => {
+    if (isDrawerOpen && menuState?.activeNavItem?.parents.includes(id)) {
+      menuState.expandNavItem(id);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [isDrawerOpen]);
+
+  if (isVisible && !isVisible()) {
```
Comment (Berkyle) [inline src/components/navigation/SlideMenuItems.tsx:105]: this isn't used anywhere, can it be removed? I feel it would be better to instead each navItem a permissions requirement, and then check if the user meets that requirement in this component later on. for now this is just bloat

File: src/contexts/User.tsx [hunk 12]
```diff
@@ -1,10 +1,11 @@
 import React, { createContext, useContext, useEffect, useMemo, useReducer } from 'react';
 
-import UserReducer, { UserAction } from '../reducers/User.Reducer';
-import useFirebaseAuth from '../hooks/useFirebaseAuth';
-import useUserCollection from '../hooks/useUserCollection';
-import { FetchedUser, User } from '../types/';
-import { logUser, usersCollection } from '../firebase/firebase';
+import { FetchedUser, User } from '@app/types';
```
Comment (Berkyle) [inline src/contexts/User.tsx:3]: honestly I was on the fence about a lot of the import plugin fluff but this dependency sorting is very nice, +1

File: src/components/navigation/SlideMenuItems.tsx (hunk 9 above)
Comment (Berkyle) [reply src/components/navigation/SlideMenuItems.tsx]: I would also appreciate some comments on this if possible, this entire component has already gotten pretty convoluted

File: src/components/navigation/SlideMenuContext.tsx (hunk 5 above)
Comment (juzerzarif) [reply src/components/navigation/SlideMenuContext.tsx:1]: That space is more for contexts that should be used in multiple places throughout the application. This context should only ever be used by the slide panel to keep centralized state so I didn't include it in there

File: src/components/navigation/SlideMenuContext.tsx (hunk 6 above)
Comment (juzerzarif) [reply src/components/navigation/SlideMenuContext.tsx:67]: Some of the items in that array are memoized callbacks that will themselves never change. I'd like to keep the context memoized since it holds centralized state for the entire nav menu and if it gets redefined on every render (which could happen outside of a state change too if a parent component re-renders) that will lead to a lot of unnecessary re-renders of components down the tree.

File: src/components/navigation/SlideMenuItems.tsx (hunk 11 above)
Comment (juzerzarif) [reply src/components/navigation/SlideMenuItems.tsx:105]: Considering how free form permissions might end up being, I'm hesitant to support an api we don't enforce. Besides, a consumer could depend on a variety of things outside of just permissions to display a nav item which is why I think a callback is the way to go. I know it's not used right now because we don't have the auth workflow fully set up but it doesn't hurt so much that it warrants being cut as bloat.

File: src/components/navigation/SlideMenuItems.tsx (hunk 9 above)
Comment (juzerzarif) [reply src/components/navigation/SlideMenuItems.tsx]: I personally like the self invoking function for scenarios like these because all of the logic is defined at the same spot you declared your variable and since it's const you know that that's the final value it's gonna have and don't have to worry about it possibly changing somewhere else in the function.

File: src/App.tsx (hunk 1 above)
Comment (juzerzarif) [reply src/App.tsx:25]: Sure https://github.com/resident51/resident51/pull/3/commits/ad9cf714da9c80a0a85a88d5891d7dee55f45ffa

File: src/components/navigation/NavigationItems.tsx (hunk 2 above)
Comment (juzerzarif) [reply src/components/navigation/NavigationItems.tsx]: I went ahead and just removed it https://github.com/resident51/resident51/pull/3/commits/ad9cf714da9c80a0a85a88d5891d7dee55f45ffa

File: src/components/navigation/SlideMenu.tsx (hunk 3 above)
Comment (juzerzarif) [reply src/components/navigation/SlideMenu.tsx]: Removed https://github.com/resident51/resident51/pull/3/commits/ad9cf714da9c80a0a85a88d5891d7dee55f45ffa

File: src/components/navigation/SlideMenu.tsx (hunk 4 above)
Comment (juzerzarif) [reply src/components/navigation/SlideMenu.tsx]: Removed https://github.com/resident51/resident51/pull/3/commits/ad9cf714da9c80a0a85a88d5891d7dee55f45ffa

File: src/components/navigation/SlideMenu.tsx (hunk 7 above)
Comment (juzerzarif) [reply src/components/navigation/SlideMenu.tsx]: It's only a warning when reported by the typescript-eslint rules. I've updated the linting config to include unused imports and setup a CI check for linting https://github.com/resident51/resident51/pull/3/commits/f379bf7c919f932e035213b5b0944921a6ffd8ad

File: src/components/navigation/SlideMenu.tsx (hunk 8 above)
Comment (juzerzarif) [reply src/components/navigation/SlideMenu.tsx]: Updated https://github.com/resident51/resident51/pull/3/commits/ad9cf714da9c80a0a85a88d5891d7dee55f45ffa

File: src/components/navigation/SlideMenuItems.tsx (hunk 9 above)
Comment (juzerzarif) [reply src/components/navigation/SlideMenuItems.tsx]: I've added some comments to explain the flow https://github.com/resident51/resident51/pull/3/commits/ad9cf714da9c80a0a85a88d5891d7dee55f45ffa

File: src/components/navigation/SlideMenuItems.tsx (hunk 10 above)
Comment (juzerzarif) [reply src/components/navigation/SlideMenuItems.tsx]: Yeah, I didn't notice that. Un-memoized https://github.com/resident51/resident51/pull/3/commits/ad9cf714da9c80a0a85a88d5891d7dee55f45ffa

File: src/components/navigation/SlideMenu.tsx (hunk 7 above)
Comment (Berkyle) [reply src/components/navigation/SlideMenu.tsx]: neat, and plus one for making our github actions actually do something

File: src/components/navigation/SlideMenuContext.tsx (hunk 6 above)
Comment (Berkyle) [reply src/components/navigation/SlideMenuContext.tsx:67]: I can live with that, better safe than sorry

File: src/components/navigation/SlideMenuItems.tsx (hunk 11 above)
Comment (Berkyle) [reply src/components/navigation/SlideMenuItems.tsx:105]: sounds good, I hadn't considered there may factors besides permissions. it doesn't hurt to keep it

File: src/components/navigation/SlideMenuContext.tsx (hunk 5 above)
Comment (Berkyle) [reply src/components/navigation/SlideMenuContext.tsx:1]: Gotcha, in that case you're right to keep it in this directory.

Just curious, but if the logic here is only used by a single component and it's direct children (SlideMenu and SlideMenuItem) why use a context instead of keeping state in the SlideMenu component? any future plans for this context?

**Review by Berkyle — APPROVED** [review]
looks good
